### PR TITLE
Prepare util::Status for Windows use

### PR DESCRIPTION
**Pull request by wilhuff**
There's a few changes in here, all related to using `util::Status` in filesystem code.

  * Moved Status into its own library to make it accessible to the different ODR-incompatible filesystem implementations
  * Moved Status::FromErrno into `status_posix.cc`
  * Added Status::FromLastError in `status_win.cc`
  * Simplified Status::FromErrno to accept format arguments and delegate to StringFormat (not sure if this is the best idea; feedback welcome)
  * Added Status::Annotate to allow us to add a description of what we were trying to do when we encountered a low-level failure

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/util/status.h [hunk 1]
```diff
@@ -93,6 +93,25 @@ class ABSL_MUST_USE_RESULT Status {
...
+  //     return s.Annotate("loading blacklist");
+  //   }
+  //
+  // Annotate() adds the appropriate separators, so callers should not include a
+  // separator in 'msg'. The exact formatting is subject to change, so you
+  // should not depend on it in your tests.
+  //
+  // OK status values have no error message and therefore if 's' is OK, the
+  // result is unchanged.
+  Status Annotate(absl::string_view msg) const;
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/util/status.h]: There's some theoretical overlap with Status::Update. The docs for that claim if the existing status is not ok, then it "preserves the current status, but may augment with additional information about `new_status`."

But it doesn't. (The word "may" being the key word there.)

Maybe it should? It could use your new Annotate to do so...

Obviously unrelated to this PR. Optional.

Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/util/status.h]: Note that Status objects aren't immutable. (Though glancing at it, the only non-const method is Update.) As such, this may be better written as `void Annotate(string_view msg); //non-const`.

Or alternatively, we could make Status immutable and alter Update instead, though that seems to diverge a bit from the other existing status implementations (if we care).

Looking at the g3 implementation, it defines this as a free function (and therefore takes the existing status as a parameter). So that's another option too. (Almost equivalently, you could also define it as a static function.)

Basically, I'm not wild about the fact that Update and Annotate treat this as mutable/immutable; they should behave similarly. Any of the above would resolve that for me.

File: Firestore/core/test/firebase/firestore/util/status_test.cc [hunk 2]
```diff
@@ -113,6 +113,18 @@ TEST(Status, FromErrno) {
           "Already exists: Cannot write file \\(errno .*: File exists\\)"));
 }
 
+TEST(Status, Annotate) {
+  Status a(FirestoreErrorCode::NotFound, "fork not found");
+  Status b = a.Annotate("could not eat dinner");
+  ASSERT_EQ(FirestoreErrorCode::NotFound, b.code());
+  ASSERT_EQ("fork not found; could not eat dinner", b.error_message());
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/util/status_test.cc]: Since you don't guarantee the separator, you might instead do:
```
ASSERT_NE(b.error_message().find("fork not found") , string::npos);
ASSERT_NE(b.error_message().find("could not eat dinner"), string::npos);
```

(Maybe gtest has some string/container "contains" helper to make this better.)

File: Firestore/core/test/firebase/firestore/util/status_test.cc [hunk 3]
```diff
@@ -113,6 +113,18 @@ TEST(Status, FromErrno) {
...
+
+  a = Status(FirestoreErrorCode::NotFound, "");
+  b = a.Annotate("could not eat dinner");
+  ASSERT_EQ(FirestoreErrorCode::NotFound, b.code());
+  ASSERT_EQ("could not eat dinner", b.error_message());
+}
+
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/util/status_test.cc]: You might also consider adding tests for:

```
TEST(Status, AnnotateOkStatusDoesntAnnotate) {
  a = Status::Ok();
  b = a.Annotate("foo");
  ASSERT_TRUE(b.ok());
  ASSERT_EQ(b.error_message().find("foo") == string::npos);
}
```

File: Firestore/core/test/firebase/firestore/util/status_test.cc (hunk 2 above)
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/util/status_test.cc]: You're asserting kinda unrelated things. Or at least, things that could fail separately for different reasons. You could change these to expect. Optional.

File: Firestore/core/src/firebase/firestore/util/status_win.cc [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+ * Returns the Canonical error code for the given Windows API error code as
+ * obtained from GetLastError().
+ */
+static FirestoreErrorCode CodeForLastError(DWORD error) {
+  switch (error) {
+    case ERROR_SUCCESS:
+      return FirestoreErrorCode::Ok;
+
+    case ERROR_INVALID_ACCESS:
+      return FirestoreErrorCode::Internal;
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/util/status_win.cc]: This sounds like it could be PermissionDenied?

https://docs.microsoft.com/en-us/windows/desktop/debug/system-error-codes--0-499- is supremely unhelpful.

"ERROR_INVALID_ACCESS

12 (0xC)

The access code is invalid."

I assume that's referring to some sort of auth token. Though maybe the reasoning here is that it isn't *our* auth token, so something else must have gone horribly wrong?

File: Firestore/core/src/firebase/firestore/util/status.h (hunk 1 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/status.h]: Agreed. I've backed this out. `CausedBy` will treat Status as mutable (to match `Update`).

File: Firestore/core/src/firebase/firestore/util/status_win.cc (hunk 4 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/status_win.cc]: Good call. I had confused this with something else.

File: Firestore/core/test/firebase/firestore/util/status_test.cc (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/status_test.cc]: There's some string matchers in gmock. I'll use those in the follow-up PR. This has been backed out.

Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/status_test.cc]: Will do in the next PR.

File: Firestore/core/test/firebase/firestore/util/status_test.cc (hunk 3 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/status_test.cc]: Will add in the next PR.

**Comment by wilhuff** [comment]
I'm going to adjust the filesystem code to just use StringFormat directly. Feel free to push back if it's too verbose.

**Comment by rsgowman** [comment]
> I'm going to adjust the filesystem code to just use StringFormat directly. Feel free to push back if it's too verbose.

sgtm.
